README.md
# Dart Sanity.io

This Dart package is a small abstraction on top of the Sanity.io API.

## Installation 

```sh
dart pub add flutter_sanity
```

## Example

```dart
// main.dart
import 'package:flutter_sanity/flutter_sanity.dart';

void main() async {
  final sanityClient = SanityClient(
    dataset: 'dataSet',
    projectId: 'projectId',
  );

  final response = await sanityClient.fetch('*[_type == "post"]');

  print(response);
}
```

Credits to: https://gist.github.com/ameistad/2c6eec79a62c7f4d08d051ef4ef14ba0

CHANGELOG.md
## 1.1.0

* Add `download` method #1

## 1.0.1

* Fix styling

## 1.0.0

* Initial release

test/src/exception_test.dart
import 'package:test/test.dart';
import 'package:flutter_sanity/src/exception.dart';

void main() {
  group('AppException', () {
    test('FetchDataException toString', () {
      expect(
        FetchDataException('fetch data').toString(),
        'Error During Communication: fetch data',
      );
    });
    test('BadRequestException toString', () {
      expect(
        BadRequestException('bad request').toString(),
        'Invalid Request: bad request',
      );
    });
    test('UnauthorizedException toString', () {
      expect(
        UnauthorizedException('unauthorized').toString(),
        'Unauthorized: unauthorized',
      );
    });
    test('InvalidInputException toString', () {
      expect(
        InvalidInputException('input').toString(),
        'Invalid Input: input',
      );
    });
    test('FormatDataException toString', () {
      expect(
        FormatDataException('format data').toString(),
        'Error During Parsing: format data',
      );
    });
  });
}


test/src/http_client_test.dart
import 'package:http/http.dart';
import 'package:http/testing.dart';
import 'package:test/test.dart';
import 'package:flutter_sanity/src/http_client.dart';

void main() {
  group('HttpClient', () {
    test('sets Authorization Header', () async {
      final client = HttpClient(
        'test',
        client: MockClient((request) async {
          expect(request.headers.containsKey('Authorization'), true);
          expect(request.headers['Authorization'], 'Bearer test');

          return Response('', 200);
        }),
      );

      await client.get(Uri(path: ''));
    });

    test('sets no Authorization Header', () async {
      final client = HttpClient(
        null,
        client: MockClient((request) async {
          expect(request.headers.containsKey('Authorization'), false);

          return Response('', 200);
        }),
      );

      await client.get(Uri(path: ''));
    });
  });
}


test/src/sanity_client_test.dart
import 'dart:convert';

import 'package:flutter_sanity/src/exception.dart';
import 'package:http/http.dart';
import 'package:http/testing.dart';
import 'package:test/test.dart';
import 'package:flutter_sanity/src/sanity_client.dart';

void main() {
  late SanityClient client;

  group('SanityClient', () {
    test('constructor', () {
      client = SanityClient(
        projectId: 'test',
        dataset: 'test',
        token: 'test',
      );

      expect(client.dataset, 'test');
      expect(client.projectId, 'test');
      expect(client.token, 'test');
      expect(client.useCdn, true);
    });

    test('responds to 200', () async {
      final mockHttpClient = MockClient((request) async {
        return Response(json.encode({'result': 'test'}), 200);
      });

      client = SanityClient(
        projectId: 'test',
        dataset: 'test',
        client: mockHttpClient,
      );

      final response = await client.fetch('');

      expect(response, 'test');
    });

    test('responds to 400', () async {
      final mockHttpClient = MockClient((request) async {
        return Response('', 400);
      });

      client = SanityClient(
        projectId: 'test',
        dataset: 'test',
        client: mockHttpClient,
      );

      expect(() async => await client.fetch(''),
          throwsA(isA<BadRequestException>()));
    });

    test('responds to 401', () async {
      final mockHttpClient = MockClient((request) async {
        return Response('', 401);
      });

      client = SanityClient(
        projectId: 'test',
        dataset: 'test',
        client: mockHttpClient,
      );

      expect(() async => await client.fetch(''),
          throwsA(isA<UnauthorizedException>()));
    });

    test('responds to 403', () async {
      final mockHttpClient = MockClient((request) async {
        return Response('', 401);
      });

      client = SanityClient(
        projectId: 'test',
        dataset: 'test',
        client: mockHttpClient,
      );

      expect(() async => await client.fetch(''),
          throwsA(isA<UnauthorizedException>()));
    });

    test('responds to 500', () async {
      final mockHttpClient = MockClient((request) async {
        return Response('', 500);
      });

      client = SanityClient(
        projectId: 'test',
        dataset: 'test',
        client: mockHttpClient,
      );

      expect(() async => await client.fetch(''),
          throwsA(isA<FetchDataException>()));
    });

    test('builds uri correctly', () async {
      final paramParameter = {'test': 'test'};
      final queryParameter = {'query': 'test'};
      final projectId = 'test';
      final dataset = 'test';

      final mockHttpClient = MockClient((request) async {
        expect(request.url.scheme, 'https');
        expect(request.url.host, '$projectId.apicdn.sanity.io');
        expect(request.url.path, '/v1/data/query/$dataset');
        expect(request.url.queryParameters,
            {...queryParameter, ...paramParameter});

        return Response(json.encode({'result': 'test'}), 200);
      });

      client = SanityClient(
        projectId: projectId,
        dataset: dataset,
        client: mockHttpClient,
      );

      await client.fetch(
        'test',
        params: paramParameter,
      );
    });

    test('builds uri correctly', () async {
      final mockHttpClient = MockClient((request) async {
        return Response('', 200);
      });

      client = SanityClient(
        projectId: 'test',
        dataset: 'test',
        client: mockHttpClient,
      );

      expect(
          () async => await client.fetch(
                'test',
              ),
          throwsA(isA<FetchDataException>()));
    });
  });
}


lib/src/exception.dart
class SanityException implements Exception {
  SanityException([
    this._message,
    this._prefix,
  ]);

  final String? _message;
  final String? _prefix;

  @override
  String toString() {
    return '$_prefix$_message';
  }
}

class FetchDataException extends SanityException {
  FetchDataException([String? message])
      : super(message, 'Error During Communication: ');
}

class BadRequestException extends SanityException {
  BadRequestException([String? message]) : super(message, 'Invalid Request: ');
}

class UnauthorizedException extends SanityException {
  UnauthorizedException([String? message]) : super(message, 'Unauthorized: ');
}

class InvalidInputException extends SanityException {
  InvalidInputException([String? message]) : super(message, 'Invalid Input: ');
}

class FormatDataException extends SanityException {
  FormatDataException([String? message])
      : super(message, 'Error During Parsing: ');
}


lib/src/sanity_client.dart
import 'dart:convert';

import 'package:flutter_sanity/src/exception.dart';
import 'package:flutter_sanity/src/http_client.dart';
import 'package:http/http.dart' as http;

class SanityClient {
  SanityClient({
    required this.projectId,
    required this.dataset,
    this.token,
    this.useCdn = true,
    http.BaseClient? client,
    this.apiVersion = 'v1',
  }) {
    _client = client ?? HttpClient(token);
  }

  /// HttpClient used to make requests
  late final http.BaseClient _client;

  /// The project ID of the Sanity.io project.
  final String projectId;

  /// The dataset of the Sanity.io project.
  final String dataset;

  /// Use the Sanity CDN to fetch data.
  final bool useCdn;

  /// The bearer token of the project to use for authentication.
  /// If not set, the client will not send the token in the header.
  final String? token;

  final String apiVersion;

  /// Builds a [Uri] for a sanity endpoint.
  Uri _buildUri(String query, {Map<String, dynamic>? params}) {
    final Map<String, dynamic> queryParameters = <String, dynamic>{
      'query': query,
      if (params != null) ...params,
    };
    return Uri(
      scheme: 'https',
      host: '$projectId.${useCdn ? 'apicdn' : 'api'}.sanity.io',
      path: '/$apiVersion/data/query/$dataset',
      queryParameters: queryParameters,
    );
  }

  /// Builds a [Uri] for a sanity endpoint.
  Uri _buildDownloadUri(String name) {
    return Uri(
      scheme: 'https',
      host: 'cdn.sanity.io',
      path: '/files/$projectId/$dataset/$name',
    );
  }

  /// Handles the response from the Sanity API.
  ///
  /// Throws a [BadRequestException], [UnauthorizedException], [FetchDataException]
  /// in case the request did not succeed
  dynamic _returnResponse(http.Response response) {
    switch (response.statusCode) {
      case 200:
        return _decodeReponse(response.body);
      case 400:
        throw BadRequestException(response.body);
      case 401:
      case 403:
        throw UnauthorizedException(response.body);
      case 500:
      default:
        throw FetchDataException(
          'Error occured while communication with server with status code: ${response.statusCode}',
        );
    }
  }

  /// Decodes the Sanity response
  ///
  /// Throws a [FetchDataException] in case decoding the response fails.
  dynamic _decodeReponse(String responseBody) {
    try {
      final responseJson = jsonDecode(responseBody);
      return responseJson['result'];
    } catch (exception) {
      throw FetchDataException('Error occured while decoding response');
    }
  }

  /// Fetches the query from the Sanity API.
  ///
  /// Throws a [SanityException] in case  request fails.
  Future<dynamic> fetch(
    String query, {
    Map<String, dynamic>? params,
    Duration? duration,
  }) async {
    final Uri uri = _buildUri(query, params: params);

    final http.Response response = duration == null
        ? await _client.get(uri)
        : await _client.get(uri).timeout(duration);

    return _returnResponse(response);
  }

  String _normalizeFileName(String ref) {
    final List<String> splitList = ref.split('-');

    return '${splitList[1]}.${splitList[2]}';
  }

  /// download file from the Sanity API.
  ///
  /// [ref] - raw file name from Sanity (file-7e79aad1cfd65cfb551dc4749eea79678384ffef-zip)
  ///
  Future<http.Response> download(String ref, {Duration? duration}) async {
    final Uri uri = _buildDownloadUri(_normalizeFileName(ref));

    return duration == null
        ? await _client.get(uri)
        : await _client.get(uri).timeout(duration);
  }
}
